### PR TITLE
Update DOM function to use wire:initial-data to find elements

### PR DESCRIPTION
**Pull request by imliam**
**TLDR**

This PR attempts to solve the problem that a lot of people are having when Livewire's JS throws an error trying to find a fingerprint that doesn't exist.

image

-----

**Context**

The `DOM.rootComponentElementsWithNoParents()` function is used by `Livewire.start()` and `Livewire.rescan()` functions to initialise components on the page.

This function finds every Livewire component on the page by searching for `[wire:id]`.

However, initialising a component actually requires the `[wire:initial-data]` attribute - as components that have _already_ been initialised have this attribute stripped away.

This has lead to the situation where lots of people are getting errors when trying to initialise components on the page, or loading sub-components in a subsequent render that were not initially present (such as inside a Blade condition or added by a loop in certain scenarios). (This has already been made partially nicer to work with when manually initialising components in this PR: https://github.com/livewire/livewire/pull/1534)

This error occurs when the component JS tries to read the initial data: https://github.com/livewire/livewire/blob/master/js/component/index.js#L28-L31

To work around this, there are solutions such as what the [Turbolinks adapter](https://github.com/livewire/turbolinks/blob/v0.1.1/src/index.js#L18-L26) does to add this `[wire:initial-data]` attribute back before re-initialising them.

---

**Solution**

This PR just changes the `DOM.rootComponentElementsWithNoParents()` function to only search for elements with `[wire:initial-data]` in the first place, which gets around this and means rescanning the DOM doesn't affect already Livewire components that have already been initialised.

It seems like these are the only changes needed to make it work - it's solved the issues occurring in my codebase and everything else appears to work as intended.

### CODE REVIEW

**Comment by calebporzio** [comment]
Thanks Liam

**Comment by sytheveenje** [comment]
Thanks @imliam, I had the same issue. I upgraded to the latest version of Livewire, and the error changed. Now it's falling over line 103.

```
message.response.serverMemo = Object.assign({}, this.serverMemo)
```

My error:
```
index.js:103 Uncaught (in promise) TypeError: Cannot read property 'fingerprint' of null
    at new Component (index.js:103)
    at onNodeAdded (index.js:579)
    at callHook (morphdom.js:117)
    at handleNodeAdded (morphdom.js:257)
    at morphdom.js:443
    at morphEl (morphdom.js:482)
    at morphdom.js:169
    at morphEl (morphdom.js:482)
    at morphdom.js:169
    at morphEl (morphdom.js:482)
Component @ index.js:84
onNodeAdded @ index.js:84
callHook @ index.js:84
handleNodeAdded @ index.js:84
(anonymous) @ index.js:84
morphEl @ index.js:84
(anonymous) @ index.js:84
morphEl @ index.js:84
(anonymous) @ index.js:84
morphEl @ index.js:84
(anonymous) @ index.js:84
morphEl @ index.js:84
(anonymous) @ index.js:84
value @ index.js:84
value @ index.js:84
value @ index.js:84
value @ index.js:84
(anonymous) @ index.js:84
Promise.then (async)
(anonymous) @ index.js:84
Promise.then (async)
value @ index.js:84
sendMessage @ index.js:84
value @ index.js:84
later @ livewire.js?id=09c5bbff75d74521e6c7:1
setTimeout (async)
(anonymous) @ livewire.js?id=09c5bbff75d74521e6c7:1
value @ index.js:84
(anonymous) @ index.js:84
value @ index.js:84
(anonymous) @ index.js:84
```

**Comment by nuernbergerA** [comment]
Hey, @sytheveenje 
please help us to keep an overview and open a new issue that references this PR, instead of posting it here.

Cheers Adrian

**Comment by sytheveenje** [comment]
Will do, no problem!
